File: typescript/test/useCase/OrderCreationUseCase.test.ts
import TestOrderRepository from "../doubles/TestOrderRepository";
import Category from "../../src/domain/Category";
import InMemoryProductCatalog from "../doubles/InMemoryProductCatalog";
import Product from "../../src/domain/Product";
import OrderCreationUseCase from "../../src/useCase/OrderCreationUseCase";
import SellItemRequest from "../../src/useCase/SellItemRequest";
import SellItemsRequest from "../../src/useCase/SellItemsRequest";
import UnknownProductError from "../../src/useCase/UnknownProductError";
import OrderStatus from "../../src/domain/OrderStatus";
import bigDecimal from "js-big-decimal";

describe('OrderCreationUseCase should', () => {
    let orderRepository: TestOrderRepository;
    let useCase: OrderCreationUseCase;

    beforeEach(() => {
        orderRepository = new TestOrderRepository();

        let food = new Category("food", new bigDecimal("10"));
        useCase = new OrderCreationUseCase(
            orderRepository,
            new InMemoryProductCatalog([
                new Product("salad", new bigDecimal("3.56"), food),
                new Product("tomato", new bigDecimal("4.65"), food)
            ])
        );
    });

    test('sell multiple items', () => {
        let request = new SellItemsRequest();
        request.requests = [
            new SellItemRequest("salad", 2),
            new SellItemRequest("tomato", 3)
        ];

        useCase.run(request);

        const insertedOrder = orderRepository.getSavedOrder();
        expect(insertedOrder?.status).toBe(OrderStatus.CREATED);
        expect(insertedOrder?.total).toEqual(new bigDecimal("23.17"));
        expect(insertedOrder?.tax).toEqual(new bigDecimal("2.10"));
        expect(insertedOrder?.currency).toBe("EUR");
        expect(insertedOrder?.items.length).toBe(2);
        expect(insertedOrder?.items[0].product.name).toBe("salad");
        expect(insertedOrder?.items[0].product.price).toEqual(new bigDecimal("3.56"));
        expect(insertedOrder?.items[0].quantity).toBe(2);
        expect(insertedOrder?.items[0].taxedAmount).toEqual(new bigDecimal("7.84"));
        expect(insertedOrder?.items[0].tax).toEqual(new bigDecimal("0.72"));
        expect(insertedOrder?.items[1].product.name).toEqual("tomato");
        expect(insertedOrder?.items[1].product.price).toEqual(new bigDecimal("4.65"));
        expect(insertedOrder?.items[1].quantity).toBe(3);
        expect(insertedOrder?.items[1].taxedAmount).toEqual(new bigDecimal("15.33"));
        expect(insertedOrder?.items[1].tax).toEqual(new bigDecimal("1.38"));
    });

    test('unknown product', () => {
        let request = new SellItemsRequest();
        request.requests.push(new SellItemRequest("unknown product", 2));

        expect(() => {
            useCase.run(request)
        }).toThrowError(UnknownProductError);
    });
});

File: csharp/OrderDispatchKata.Tests/UseCase/OrderCreationUseCaseTest.cs
using System.Collections.Generic;
using System.Linq;
using NUnit.Framework;
using OrderDispatchKata.Domain;
using OrderDispatchKata.Repository;
using OrderDispatchKata.Tests.Doubles;
using OrderDispatchKata.UseCase;

namespace OrderDispatchKata.Tests.UseCase;

[TestFixture]
public class OrderCreationUseCaseTest
{
    [SetUp]
    public void SetUp()
    {
        orderRepository = new TestOrderRepository();
        food = new Category();
        food.setName("food");
        food.setTaxPercentage(10m);

        var salad = new Product();
        salad.setName("salad");
        salad.setPrice(3.56m);
        salad.setCategory(food);

        var tomato = new Product();
        tomato.setName("tomato");
        tomato.setPrice(4.65m);
        tomato.setCategory(food);

        productCatalog = new InMemoryProductCatalog(
            new List<Product>
            {
                salad, tomato
            });

        useCase = new OrderCreationUseCase(orderRepository, productCatalog);
    }

    private TestOrderRepository orderRepository;
    private Category food;
    private ProductCatalog productCatalog;
    private OrderCreationUseCase useCase;

    [Test]
    public void sellMultipleItems()
    {
        var saladRequest = new SellItemRequest();
        saladRequest.setProductName("salad");
        saladRequest.setQuantity(2);

        var tomatoRequest = new SellItemRequest();
        tomatoRequest.setProductName("tomato");
        tomatoRequest.setQuantity(3);

        var request = new SellItemsRequest();
        request.setRequests(new List<SellItemRequest>());
        request.getRequests().Add(saladRequest);
        request.getRequests().Add(tomatoRequest);

        useCase.run(request);

        var insertedOrder = orderRepository.getSavedOrder()!;
        Assert.That(insertedOrder.getStatus(), Is.EqualTo(OrderStatus.CREATED));
        Assert.That(insertedOrder.getTotal(), Is.EqualTo(23.20m));
        Assert.That(insertedOrder.getTax(), Is.EqualTo(2.13m));
        Assert.That(insertedOrder.getCurrency(), Is.EqualTo("EUR"));
        Assert.That(insertedOrder.getItems()!.Count(), Is.EqualTo(2));
        Assert.That(insertedOrder.getItems()![0].getProduct()!.getName(), Is.EqualTo("salad"));
        Assert.That(insertedOrder.getItems()![0].getProduct()!.getPrice(), Is.EqualTo(3.56m));
        Assert.That(insertedOrder.getItems()![0].getQuantity(), Is.EqualTo(2));
        Assert.That(insertedOrder.getItems()![0].getTaxedAmount(), Is.EqualTo(7.84m));
        Assert.That(insertedOrder.getItems()![0].getTax(), Is.EqualTo(0.72m));
        Assert.That(insertedOrder.getItems()![1].getProduct()!.getName(), Is.EqualTo("tomato"));
        Assert.That(insertedOrder.getItems()![1].getProduct()!.getPrice(), Is.EqualTo(4.65m));
        Assert.That(insertedOrder.getItems()![1].getQuantity(), Is.EqualTo(3));
        Assert.That(insertedOrder.getItems()![1].getTaxedAmount(), Is.EqualTo(15.36m));
        Assert.That(insertedOrder.getItems()![1].getTax(), Is.EqualTo(1.41m));
    }

    [Test]
    public void unknownProduct()
    {
        var request = new SellItemsRequest();
        request.setRequests(new List<SellItemRequest>());
        var unknownProductRequest = new SellItemRequest();
        unknownProductRequest.setProductName("unknown product");
        request.getRequests().Add(unknownProductRequest);

        Assert.That(() => useCase.run(request),
            Throws.TypeOf<UnknownProductException>());
    }
}